src/devices/portaudioRecorder/PortAudioRecorderDeviceDriver - Copy.cpp

```diff
@@ -0,0 +1,395 @@
+typedef char SAMPLE;
+#define SAMPLE_SILENCE  (0)
+#else
+#define PA_SAMPLE_TYPE  paUInt8
+typedef unsigned char SAMPLE;
+#define SAMPLE_SILENCE  (128)
+#define SAMPLE_UNSIGNED
+#endif
+
+/* This routine will be called by the PortAudio engine when audio is needed.
+** It may be called at interrupt level on some machines so don't do anything
+** that could mess up the system like calling malloc() or free().
+*/
+static int bufferIOCallback( const void *inputBuffer, void *outputBuffer,
+                         unsigned long framesPerBuffer,
+                         const PaStreamCallbackTimeInfo* timeInfo,
+                         PaStreamCallbackFlags statusFlags,
+                         void *userData )
+{
+    auto* dataBuffers = static_cast<circularDataBuffers*>(userData);
+    CircularAudioBuffer_16t *recdata  = dataBuffers->recData;
+    int num_rec_channels     = dataBuffers->numRecChannels;
+    int finished = paComplete;
+
+    if (dataBuffers->canRec)
+    {
+        const auto* rptr = (const SAMPLE*)inputBuffer;
+        unsigned int framesToCalc;
+        unsigned int i;
+        size_t framesLeft = (recdata->getMaxSize().getSamples()* recdata->getMaxSize().getChannels()) -
+                            (recdata->size().getSamples()      * recdata->size().getChannels());
+
+        (void) outputBuffer; // just to prevent unused variable warnings
+        (void) timeInfo;
+        (void) statusFlags;
+        (void) userData;
+
+        if( framesLeft/ num_rec_channels < framesPerBuffer )
+        {
+            framesToCalc = framesLeft/ num_rec_channels;
+#ifdef STOP_REC_ON_EMPTY_BUFFER
+            //if we return paComplete, then the callback is not called anymore.
+            //method Pa_IsStreamActive() will return 1.
+            //user needs to call Pa_StopStream() before starting a new recording session
+            finished = paComplete;
+#else
+            finished = paContinue;
+#endif
+        }
+        else
+        {
+            framesToCalc = framesPerBuffer;
+            //if we return paContinue, then the callback will be invoked again later
+            //method Pa_IsStreamActive() will return 0
+            finished = paContinue;
+        }
+
+        if( inputBuffer == nullptr )
+        {
+            for( i=0; i<framesToCalc; i++ )
+            {
+                recdata->write(0); // left
+                if(num_rec_channels == 2 ) recdata->write(0);  // right
+            }
+        }
+        else
+        {
+#if 0
+            yDebug() << "Writing" << framesToCalc*2*2 << "bytes in the circular buffer";
+#endif 
+            for( i=0; i<framesToCalc; i++ )
+            {
+                recdata->write(*rptr++);  // left
+                if(num_rec_channels == 2 ) recdata->write(*rptr++);  // right
+            }
+        }
+        return finished;
+    }
+
+    printf("No write operations requested, aborting\n");
+    return paAbort;
+}
+
+PortAudioRecorderDeviceDriver::PortAudioRecorderDeviceDriver() :
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Member variable 'PortAudioRecorderDeviceDriver::m_isRecording' is not initialized in the constructor.](https://app.codacy.com/manual/robotology/yarp/pullRequest?prid=5162834)